[WIP] Beginnings of Trello Bot

blomquisg opened the pull request:
[Trello Bot (Issue #95)](https://github.com/ManageIQ/miq_bot/issues/95) is a bot that will help us manage our usage of Trello in a few different ways:

1. Link bugs to Trello

  By linking bugs into Trello, developers will be able to see exactly which bugs they're responsible for in the same context as the feature work that they're responsible for.  In other words, it puts all of the work in one place.

  Also, it will give non-developers visibility into the **true** velocity of the team by incorporating the bugs that were fixed during a sprint as well as the feature work.

  As a bonus, developers will be able to check off bugs in card checklists to indicate that the bugs should be moved from "ASSIGNED" to "ON_DEV", or "ON_DEV" to "POST" directly from Trello without having to update Bugzilla manually.

2. Link Github to Trello

  As trello cards are mentioned in Github issues and PRs, trello cards can be updated with backlinks to those issues and PRs.  This improves traceability between Trello and Github.

3. Summary rollup boards

  Trello bot will also be able to create Rollup Boards to show over all objective summaries.  These rollups will allow non-developers to see the high-level picture of the progress of features across the workloads represented in Trello, while allowing developers to stay focused on the detailed views of their own Trello boards.

blomquisg commented:
@Fryguy, @imtayadeway, thoughts?

Fryguy commented on app/workers/trello_monitor.rb:
```diff
@@ -0,0 +1,38 @@
+  include Sidekiq::Worker
+  include Sidetiq::Schedulable
+  include MiqToolsServices::SidekiqWorkerMixin
+  sidekiq_options :queue => :trello_bot, :retry => false
```
This will have to be miq_bot in the end.

blomquisg replied:
Oh, right ... I changed that in case I wanted to run it and not accidentally run the other workers...

I'm not altogether clear on how to do that, but I thought this might be an approach?

Fryguy commented on config/settings.yml:
```diff
@@ -14,3 +14,100 @@ issue_manager:
+    developer_key:
+    member_token:
+  organization_id:
+  teams:
```
These should be defined elsewhere during deployment (I'm sure you already knew that...just clarfiying)

blomquisg replied:
The teams?  Or, the auth keys?

I sorta like having the teams in this file.  It makes it really easy to define the mappings of teams to boards and bug groups, etc... 

Maybe you mean something else?

Fryguy replied:
The teams.  I don't mind them being in configuration, just not in settings.yml.  They should be in settings.local.yml (which is not tracked) which is how we deploy the bot.

blomquisg replied:
Ah, right ...

So, I have them locally in `development.local.yml`.  I can change this to be more generic (e.g., "team1, team2", etc...).

Fryguy commented on lib/bot/trello_helper.rb:
```diff
@@ -0,0 +1,211 @@
+        )
+        ActiveBugzilla::Base.service = service
+      end
+      ActiveBugzilla
```
This is already provided.  So, no need to rewrite.

blomquisg replied:
K, I didn't dig very deep into the other helpers to find this ... I can replace this with a reference to another helper.

Fryguy commented:
I think the trello stuff should live in a service object in miq_tools_services much like the rest of the services.  At that point the general setup is pretty straightforward.

Let's pair on this a little.  This is really awesome stuff!

Fryguy commented:
Also, thinking in a parallel model, the monitor should probably just poll something, look for changes, whatever, and then kick asynchronously kick off other work items.

Fryguy commented:
A couple more things...

Don't put anything in lib/bot...that directory is on the chopping block.  It can live in lib directly, or eventually moved to miq_tools_services if that is more appropriate.

blomquisg commented:
> ...Don't put anything in lib/bot...
> ...the trello stuff should live in a service object in miq_tools_services...
> ...thinking in a parallel model, the monitor should probably just poll...

Ah, cool ... I can move that into lib for now.  Then, we can talk about what should move out to a separate service.

> Don't forget to the link the GH issue

Added to the PR summary.

blomquisg commented:
> Don't put anything in lib/bot

moved to `lib`

> [BZ service stuff] is already provided

using `MiqToolsServices::Bugzilla` now

> The teams...should be in settings.local.yml

updated `settings.yml` to just have examples

Fryguy commented on Gemfile:
```diff
@@ -61,3 +61,7 @@ group :issue_manager do
   gem 'faraday', '~> 0.9.1'
 end
+
+group :trello do
```
Not sure we need a group at all.
